driver-core/src/main/com/mongodb/internal/connection/DnsMultiServerCluster.java

```diff
@@ -57,17 +59,35 @@ public void initialize(final Collection<ServerAddress> hosts) {
                 }
             }
 
-            private Collection<ServerAddress> applySrvMaxHosts(final Collection<ServerAddress> hosts) {
-                Collection<ServerAddress> newHosts = hosts;
+            private Collection<ServerAddress> applySrvMaxHosts(final Collection<ServerAddress> latestSrvHosts) {
                 Integer srvMaxHosts = getSettings().getSrvMaxHosts();
-                if (srvMaxHosts != null && srvMaxHosts > 0) {
-                    if (srvMaxHosts < hosts.size()) {
-                        List<ServerAddress> newHostsList = new ArrayList<>(hosts);
-                        Collections.shuffle(newHostsList, ThreadLocalRandom.current());
-                        newHosts = newHostsList.subList(0, srvMaxHosts);
-                    }
+                if (srvMaxHosts == null || srvMaxHosts <= 0 || latestSrvHosts.size() <= srvMaxHosts) {
+                    return new ArrayList<>(latestSrvHosts);
                 }
-                return newHosts;
+                List<ServerAddress> activePriorHosts = getActivePriorHosts(latestSrvHosts);
+                int numNewHostsToAdd = srvMaxHosts - activePriorHosts.size();
+                List<ServerAddress> result = getShuffledLatestSrvHosts(latestSrvHosts, activePriorHosts, numNewHostsToAdd);
+
+                return result;
```

Comment:
This is cleaner than the original, and the naming (active, latest, etc.) is clear. Nice!